naga/src/back/spv/index.rs

```diff
@@ -38,98 +38,160 @@ impl<'w> BlockContext<'w> {
+    /// The SPIR-V generated takes into account wrapped globals; see
+    /// [`global_needs_wrapper`].
+    ///
+    /// [`GlobalVariable`]: crate::Expression::GlobalVariable
+    /// [`AccessIndex`]: crate::Expression::AccessIndex
+    /// [`Access`]: crate::Expression::Access
+    /// [`base`]: crate::Expression::Access::base
     pub(super) fn write_runtime_array_length(
         &mut self,
         array: Handle<crate::Expression>,
         block: &mut Block,
     ) -> Result<Word, Error> {
-        // Naga IR permits runtime-sized arrays as global variables, or as the
-        // final member of a struct that is a global variable, or one of these
-        // inside a buffer that is itself an element in a buffer bindings array.
-        // SPIR-V requires that runtime-sized arrays are wrapped in structs.
-        // See `helpers::global_needs_wrapper` and its uses.
-        let (opt_array_index_id, global_handle, opt_last_member_index) = match self
-            .ir_function
-            .expressions[array]
-        {
+        // The index into the binding array, if any.
+        let binding_array_index_id: Option<Word>;
+
+        // The handle to the Naga IR global we're referring to.
+        let global_handle: Handle<crate::GlobalVariable>;
+
+        // At the Naga type level, if the runtime-sized array is the final member of a
+        // struct, this is that member's index.
+        //
+        // This does not cover wrappers: if this backend wrapped the Naga global's
+        // type in a synthetic SPIR-V struct (see `global_needs_wrapper`), this is
+        // `None`.
+        let opt_last_member_index: Option<u32>;
+
+        // Inspect `array` and decide whether we have a binding array and/or an
+        // enclosing struct.
+        match self.ir_function.expressions[array] {
             crate::Expression::AccessIndex { base, index } => {
                 match self.ir_function.expressions[base] {
-                    // The global variable is an array of buffer bindings of structs,
-                    // we are accessing one of them with a static index,
-                    // and the last member of it.
                     crate::Expression::AccessIndex {
                         base: base_outer,
                         index: index_outer,
                     } => match self.ir_function.expressions[base_outer] {
+                        // An `AccessIndex` of an `AccessIndex` must be a
+                        // binding array holding structs whose last members are
+                        // runtime-sized arrays.
                         crate::Expression::GlobalVariable(handle) => {
                             let index_id = self.get_index_constant(index_outer);
-                            (Some(index_id), handle, Some(index))
+                            binding_array_index_id = Some(index_id);
+                            global_handle = handle;
+                            opt_last_member_index = Some(index);
```

Comment:
nitpick(non-blocking): No action required—I understand what I'm looking at now—but it woulda been nice if the change from `let (…) = …` to `let $ident; …` had been kept separate from the documentation and error content changes.

Reply:
Yes, this could have been separated out better. Everything was written at once, and I was afraid I would introduce bugs or confuse myself trying to reconstitute them as a series of individual commits. Some of the changes invalidated some of the old comments (variable renames), so the code and doc changes couldn't really be cleanly separated out, line by line. Of course, I could have just rewritten the change from scratch, comparing with the final result, but I felt that had crossed the line of diminishing returns.